rfd/0133-connect-my-computer.md

```diff
@@ -0,0 +1,584 @@
+
+#### Downloading the agent
+
+
+To download the teleport binary, Connect needs to construct a URL of the tar.gz file with binaries.
+The URL contains the OS, the architecture and the agent version. The OS and the architecture are
+already kept in `RuntimeSettings`.
+
+For Linux, the downloaded file is one of:
+* `teleport-v<version>-linux-arm64-bin.tar.gz`
+   * We don’t have Connect arm64 releases, but it will be useful for testing on VMs.
+* `teleport-v<version>-linux-amd64-bin.tar.gz`
+
+For macOS:
+* `teleport-v<version>-darwin-arm64-bin.tar.gz`
+* `teleport-v<version>-darwin-amd64-bin.tar.gz`
+
+The MVP will not verify the checksum of the downloaded archive.
+
+##### Agent versioning
+
+The version of the agent that Connect uses always matches the version of Connect. If Connect detects
+that the combination of the agent version and the cluster version breaks [the compatibility promise](https://goteleport.com/docs/faq/#version-compatibility),
+it shows a warning on the setup page and the status page, informing about lack of support and
+recommending an upgrade or a downgrade.
+
+As such, CMC is considered to be supported only when Connect is on the same major version or one
+major version behind the cluster.
+
+To check if the versions are compatible, Connect compares its version with the cluster version from
+the `Ping` response.
+
+This is how the UI behaves depending on the version differences:
+
+* If the agent is on the same version as the cluster, everything's okay and Connect doesn't show any
+  extra information.
+* If the agent is two major versions or more _behind_ the cluster, Connect shows a warning on the
+  setup page and the status page explaining that servers do not support clients that are more than
+  one major version behind, links to `https://goteleport.com/docs/faq/#version-compatibility` and
+  recommends _an upgrade_ of Connect.
+   * Examples: Connect v11.0.0 & cluster v13.1.1, Connect v10.3.16 & cluster v12.0.1.
+* If the agent is one major version or more _ahead of_ the cluster, Connect shows a warning on the
+  setup page and the status page explaining that servers do not support clients that are on a newer
+  major version, links to `https://goteleport.com/docs/faq/#version-compatibility` and recommends _a
+  downgrade_ of Connect.
+   * Examples: Connect v14.0.0 & cluster v13.1.1, Connect v10.3.16 & cluster v9.3.26.
+* If the agent is on a supported client version that is older than the cluster version, Connect
+  shows an informational message only on the status page recommending an upgrade, similar to [how
+  the Web UI does it](https://github.com/gravitational/teleport/blob/12204708e71704a8073a6ee5d2b2358db23071c1/lib/auth/auth.go#L1101-L1103),
+  and links to the download page.
+   * Examples: Connect v13.1.0 & cluster v13.1.1, Connect v12.0.0 & cluster v13.1.0.
+
+It's important that the warnings are shown indeed as warnings and not as errors – we don't want to
```

Comment:
I'm not sure about this. These are likely new users, so preventing them from running an unsupported mix of versions might be a good thing (otherwise they might see confusing error messages or other unexpected behavior and end up not liking Teleport as a result). Simply refusing to run an incompatible agent might result in stronger first impressions.

Reply:
Now that I think about it, I kind of agree.

Preventing the user from running an unsupported mix of versions is something we initially wanted to do. However, we realized that:

1) It takes Cloud usually at least a month to deploy a new major version. Say we release it in v14, someone reads about it and decides to check it out. They sign up for the Team plan, download Connect and open the setup page. At this point all we'd say is "Whoops, you see you downloaded the newest version of Connect, but Cloud is actually a major version behind, so you have to go back and download V-1".
2) Most of the time [citation needed] new agent versions work with old clusters… probably?

Now, the counterarguments for those points could be:

1)
   *  Well, we could just fix the download page so that the Discover tile points at the download page with the correct version of Connect being chosen.
      * This doesn't really address an issue where someone learns about this feature outside of Connect, signs up for the Team plan and then tries to use CMC.
         * We could argue that this won't happen very often and if it does happen, downloading an older version is not a huge cost (especially if we were able to directly navigate the user towards that older version).
   * We plan to completely change how Cloud adopts new versions. Besides, the original issue would be a problem only roughly for 30 days after a new release.
2) Just because a v13 agent works with a v12 cluster, or v14 with v13, doesn't mean that it'll continue to do so. Matter of fact, this explicitly goes against our compatibility promise.

Now I'm partial to disallowing unsupported versions outright. This shouldn't be a huge cost for the users to pay, especially if we can link them directly to a supported version. So we could have a link like `https://goteleport.com/download/?v=12` that shows 12.4.8.

Reply:
> new agent versions work with old clusters

This is explicitly **against** our compatibility guarantees and likely to fail.

> It takes Cloud usually at least a month to deploy a new major version.

This will not be the case for much longer.

Given the above, I agree with you - let's just disallow unsupported versions.